xdsl/traits.py

```diff
@@ -444,30 +445,98 @@ def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
+    The kind of side effect an operation can have.
+
+    MLIR has a more detailed version of this, able to tie effects to specfic resources or
+    values. Here, everything has its effect on the universe.
+    """
+
+    READ = auto()
+    """
+    Indicates that the operation reads from some resource. A 'read' effect implies only
+    dereferencing of the resource, and not any visible mutation.
+    """
+
+    WRITE = auto()
+    """
+    Indicates that the operation writes to some resource. A 'write' effect implies only
+    mutating a resource, and not any visible dereference or read.
+    """
+
+    ALLOC = auto()
+    """
+    Indicates that the operation allocates from some resource. An 'allocate' effect
+    implies only allocation of the resource, and not any visible mutation or dereference.
+    """
+
+    FREE = auto()
+    """
+    Indicates that the operation frees some resource that has been allocated. A 'free'
+    effect implies only de-allocation of the resource, and not any visible allocation,
+    mutation or dereference.
+    """
+
+
 class MemoryEffect(OpTrait):
     """
     A trait that enables operations to expose their side-effects or absence thereof.
-
-    NB: The MLIR implementation further allows to describe what *kind* of side-effects
-    an operation has, e.g., read-only, or allocation.
-    This one is a stripped down version for now, just saying if there are any
-    side-effects or not.
     """
 
     @classmethod
     @abc.abstractmethod
-    def has_effects(cls, op: Operation) -> bool:
+    def get_effects(cls, op: Operation) -> set[MemoryEffectKind] | None:
+        """
+        Returns the concrete side effects of the operation.
+
+        Return None if the operation cannot conclude - interpreted as if the operation
+        had no MemoryEffect interface in the first place.
+        """
         raise NotImplementedError()
 
+    @final
+    @classmethod
+    def has_effects(cls, op: Operation) -> bool:
+        """
+        Returns if the operation has any side effects.
+        """
+        effects = cls.get_effects(op)
+        return (effects is not None) and len(effects) > 0
+
 
-def is_side_effect_free(op: Operation):
+def only_has_effect(op: Operation, effect: MemoryEffectKind) -> bool:
+    """
+    Returns if the operation has the given side effects and no others.
+    """
+    return get_effects(op) == {effect}
+
+
+def is_side_effect_free(op: Operation) -> bool:
     """
     Boilerplate helper to check if a generic operation is side effect free for sure.
     """
-    # If it doesn't say, safely assume it has side effects.
-    if not (trait := op.get_trait(MemoryEffect)):
-        return False
-    return not trait.has_effects(op)
+    return get_effects(op) == set()
+
+
+def get_effects(op: Operation) -> set[MemoryEffectKind] | None:
+    """
+    Helper to get known side effects of an operation, including recursive effects.
+    None means that the operation has unknown effects, for safety.
```

Comment:
I think None here is a bit weird. would it make sense to add an `unknown` to the memory effect kinds? IMO, this would be much clearer and you could also express that something is guaranteed to have a write effect and maybe others and that something is guaranteed to have a write effect and no others. Currently, this distinction is not possible, AFAIU.

Reply:
but I guess we are copying MLIR here so....

Reply:
Actually not on this one ! This is an extension from MLIR already, which only allows to return known effects or not implement the interface at all.

This is a minimal extension code wise, and already allows unifying RecursiveEffect, which otherwise has to be a separate trait (as it could, well, have recursively unknown effects)

If you're fine with it, I would like to keep this probably non-final extension as-is; I am just after porting MLIR functionality here initially, not improve the design fundamentally, if that makes sense

I want to be clear, I think there's a lot of interesting discussions to have on those nuances and not dismissing that. I'd just like to push it out of this initial xDSL PR.

Reply:
Okay, sure, I am fine with leaving it as-is :+1: